[DOC-7830] Add more details about finalization, migrations, and how to monitor progress (#18609)
cockroachdb/docs: src/current/v22.2/upgrade-cockroach-version.md

1. Restart the `cockroach` process on the node and verify that it has rejoined the cluster before rolling back the upgrade on the next node.
1. After all nodes have been rolled back and rejoined the cluster, [finalize the rollback](#step-6-finish-the-upgrade) in the same way as you would finalize an upgrade, as described in the next section.

> Because a finalized major-version upgrade cannot be rolled back, Cockroach Labs recommends that you monitor the stability and performance of your cluster with the upgraded binary for at least a day before deciding to finalize the upgrade.

Author: ```suggestion
## Step 6. Finish the upgrade

Because a finalized major-version upgrade cannot be rolled back, Cockroach Labs recommends that you monitor the stability and performance of your cluster with the upgraded binary for at least a day before deciding to finalize the upgrade.
```

## Step 6. Finish the upgrade

{{site.data.alerts.callout_info}}
Finalization is required only when upgrading from {{ previous_version }}.x to {{ page.version.version }}. For upgrades within the {{ page.version.version }}.x series, skip this step.
{{site.data.alerts.end}}

> If you disabled auto-finalization in [step 3](#step-3-decide-how-the-upgrade-will-be-finalized), monitor the stability and performance of your cluster for as long as you require to feel comfortable with the upgrade (generally at least a day). If during this time you decide to roll back the upgrade, repeat the rolling restart procedure with the previous binary. Otherwise, you must re-enable upgrade finalization to complete the upgrade to {{ page.major_version }}. Cockroach Labs recommends that you either finalize or roll back a major-version upgrade within a relative short period of time; running in a partially-upgraded state is not recommended.

Reviewer: ```suggestion
If you disabled auto-finalization in [step 3](#step-3-decide-how-the-upgrade-will-be-finalized), monitor the stability and performance of your cluster for at least a day. If you decide to roll back the upgrade, repeat the [rolling restart procedure](#step-4-perform-the-rolling-upgrade) with the previous binary. Otherwise, you must re-enable upgrade finalization to complete the upgrade to {{ page.version.version }}. Cockroach Labs recommends that you either finalize or roll back a major-version upgrade within a relative short period of time; running in a partially-upgraded state is not recommended.
```

{{site.data.alerts.callout_danger}}

> A cluster that is not finalized on {{ previous_version }} cannot be upgraded to {{ page.major_version }} until the {{ previous_version }} upgrade is finalized.

Reviewer: Isn't this a tautology?

```suggestion
A cluster that is not finalized on {{ previous_version }} cannot be upgraded to {{ page.version.version }} until the {{ previous_version }} upgrade is finalized.
```

Author: No, customers forget to finalize after a major-version upgrade all the time. There's nothing in the UI or logs that reminds them that it's not finalized, afaik. If they never finalized an upgrade to v23.2, for example, they won't even be able to attempt to upgrade to v24.1 until they have finalized on v23.2, even if they have installed several v23.2 patches.

{{site.data.alerts.end}}

> When finalization begins, a series of migration jobs run to enable certain types of features and changes in the new major version that cannot be rolled back. These include changes to system schemas, indexes, and descriptors, and enabling certain types of improvements and new features. Until the upgrade is finalized, these features and functions will not be available and the command `SHOW CLUSTER SETTING version` will return `{{ previous_version }}`.

Reviewer: This is displaying `v23.2` in preview. But in your cockroachcloud doc edit, it seems it should display `23.2`. Which is correct?

Author: Yes, good catch.

{% capture previous_version_numeric %}{{ previous_version | strip_first: 'v'}}{% endcapture %}
{% capture major_version_numeric %}{{ page.major_version | strip_first: 'v' }}{% endcapture %}

> You can monitor the process of the migration in the DB Console [Jobs page]({% link {{ page.version.version }}/ui-jobs-page.md %}). Migration jobs have names in the format `{{ major_version_numeric }}-{migration-id}`. If a migration job fails or stalls, Cockroach Labs can use the migration ID to help diagnose and troubleshoot the problem. Each major version has different migration jobs with different IDs.

Reviewer: Is it possible to integrate this and the paragraphs below with the numbered steps? I think it is confusing to have several paragraphs of prescriptive text preceding another series of numbered steps. It looks like these paragraphs would apply between steps 2 and 3.

Author: Actually, after you finish step 4, you have to make an active decision that puts you into step 5 (Roll back the upgrade) or step 6 (Finish the upgrade). Otherwise you are stuck in the unfinalized state where you can't access the new features or do a subsequent major-version upgrade. If you are doing step 6, the only way to know that finalization has finished is to query for the version, and the only way to know its progress if it hasn't yet finished is to do line 273. Does that make sense?

Author: Good idea!

Finalization is complete when all migration jobs have completed. After migration is complete, the command `SHOW CLUSTER SETTING version` will return `{{ major_version_numeric }}.

{{site.data.alerts.callout_info}}

> All [schema change]({% link {{ page.version.version }}/online-schema-changes.md %}) jobs must reach a terminal state before finalization can complete. Finalization can therefore take as long as the longest-running schema change. Otherwise, the amount of time required for finalization depends on the amount of data in the cluster, as it kicks off various internal maintenance and migration tasks. During this time, the cluster will experience a small amount of additional load.

Reviewer: ```suggestion
All [schema change]({% link {{ page.version.version }}/online-schema-changes.md %}) jobs must reach a terminal state before finalization can complete. Finalization can therefore take as long as the longest-running schema change. Otherwise, the amount of time required for finalization depends on the amount of data in the cluster, as the process runs various internal maintenance and migration tasks. During this time, the cluster will experience a small amount of additional load.
```

    > RESET CLUSTER SETTING cluster.preserve_downgrade_option;
    ~~~

>     The series of migration jobs for {{ page.major_version }} runs in sequence.

Reviewer: The `major_version` variable isn't displaying anything in preview.

```suggestion
    The series of migration jobs for {{ page.version.version }} runs in sequence.
```

Author: I fixed the bug, it is working now.
